File: src/keys/genKeys.ts
import { generatePrivate as generatePrivateECDSA, getPublic as getPublicECDSA } from "eccrypto";
import { toKeyMap } from '../utility';
import forge from "node-forge";


const format = "pem"
const curve = "secp256k1"
/**
 * Generates equivalent to RSASSA-PKCS1-v1_5 keypair
 */
function generateKeyPairRSA() {
   
      var keypair = forge.pki.rsa.generateKeyPair(2048, 0x10001);
      return ({
        privateKey: forge.pki.privateKeyToPem(keypair.privateKey),
        publicKey:  forge.pki.publicKeyToPem(keypair.publicKey)
      })
  }

/**
 * build an ECDSA keypair using the sect571r1 curve. These keys will be used to sign and verify the authenticity of messages by the `ConnectionWrapper`
 * @returns a public and a private key.
 */
function generateKeyPairECDSA(): Promise<{
  publicKey: string,
  privateKey: string,
}> {
  return new Promise((resolve, reject) => {
    const priv = generatePrivateECDSA();
    const pub = getPublicECDSA(priv);

    resolve({
      privateKey: priv.toString("base64"),
      publicKey: pub.toString("base64")
    })
  });
}

/**
 * a key signed by any number of CAs. This is used for RSA and ECDSA by the PublicClientKey
 */
interface SignedKey {
  /**
   * The data containing the actual Key
   */
  key: string,
  /**
   * the signatures of this key. (currently unused)
   */
  signatures: Record<string, string>
}

/**
 * A clients public key. Used to authenticate that client, sign messages and communicate the chat keys. (the aes keys used to encript messages in a certain chat)
 */
interface PublicClientKey {
  /**
   * the Clients Public RSA key
   */
  data_key: SignedKey,
  /**
   * the clients Public ECDSA key
   */
  sign_key: SignedKey,
  /**
   * the user id of the client
   */
  id: string
}

/**
 * A clients private key used to create a connection to the server and handle the client side 
 */
interface PrivateClientKey {
  /**
   * the private ECDSA key
   */
  sign_key: string,
  /**
   * the privat RSA Key
   */
  data_key: string,
  /**
   * the ECDSA key of the server (currently unused)
   */
  server_key: string,
  /**
   * the public keys of the CA's (currntly unused)
   */
  ca_keys: Record<string, string>,
  /**
   * the user id of the user.
   */
  uid: string,
}

/**
 * generates the private and public keys for a certain client. The private key should be stored for later use, while the public key should be put ont the servers database and signed with the CA keys.
 * @param server_key the public key of the server. used to verify the authenticity of messages
 * @param ca_keys the public keys of the CA's used to verify the authenticity of keys recieved fromt the server.
 * @param id the users it.
 * @returns the public and private keys of the client.
 */
async function generateClientKeys(server_key: string, ca_keys: any[], id: string) {
  const { privateKey, publicKey } = await generateKeyPairECDSA();
  const { privateKey: priv_data, publicKey: pub_data } = await generateKeyPairRSA();
  return {
    privateKey: {
      sign_key: privateKey,
      data_key: priv_data,
      server_key,
      ca_keys: toKeyMap(ca_keys, "id"),
      uid: id,
    } as PrivateClientKey,
    publicKey: {
      data_key: {
        key: pub_data,
        signatures: {}
      },
      sign_key: {
        key: publicKey,
        signatures: {}
      },
      id
    } as PublicClientKey
  }
}

export { generateClientKeys }
export type { PublicClientKey, PrivateClientKey, SignedKey }

File: cpp/src/keys/genKeys.cpp
#include <elonef-communication/keys/genKeys.hpp>
#include <elonef-communication/utils.hpp>
#include <cryptopp/osrng.h>
#include <cryptopp/rsa.h>
#include <cryptopp-pem/pem.h>
#include <elonef-communication/types.hpp>
#include <elonef-communication/encoding/encoding.hpp>
#include <elonef-communication/encoding/decoding.hpp>
#include <iostream>

std::string buildIndenter(size_t indent) {
    std::string indenter;
    for (size_t idx=0; idx<indent; idx++) {
        indenter = indenter + "\t";
    }
    return indenter;
}

Elonef::SignedKey::SignedKey(CryptoPP::ByteQueue& data) {
    this->key = Elonef::toDynamicSizeString(data);
    this->signatures = Elonef::toMap<std::string, std::string>(data, toDynamicSizeString_long, toDynamicSizeString_long);
}

Elonef::SignedKey::SignedKey() {}

CryptoPP::ByteQueue Elonef::SignedKey::toQueue() const {
    CryptoPP::ByteQueue out = Elonef::toBytes(this->key);
    Elonef::toBytes(signatures.begin(), signatures.end()).TransferAllTo(out);
    return out;
}

Elonef::PublicClientKey::PublicClientKey(CryptoPP::ByteQueue& data) {
    this->id = Elonef::toDynamicSizeString(data);
    this->data_key = Elonef::SignedKey(data);
    this->sign_key = Elonef::SignedKey(data);
}

Elonef::PublicClientKey::PublicClientKey() {}

CryptoPP::ByteQueue Elonef::PublicClientKey::toQueue() const {
    CryptoPP::ByteQueue out = Elonef::toBytes(this->id);
    this->data_key.toQueue().TransferAllTo(out);
    this->sign_key.toQueue().TransferAllTo(out);
    return out;
}

Elonef::PrivateClientKey::PrivateClientKey(CryptoPP::ByteQueue& data) {
    this->sign_key = Elonef::toDynamicSizeString(data);
    this->data_key = Elonef::toDynamicSizeString(data);
    this->server_key = Elonef::toDynamicSizeString(data);
    this->ca_keys = Elonef::toMap<std::string, std::string>(data, toDynamicSizeString_long, toDynamicSizeString_long);
    this->uid = Elonef::toDynamicSizeString(data);
}

Elonef::PrivateClientKey::PrivateClientKey() {}

CryptoPP::ByteQueue Elonef::PrivateClientKey::toQueue() const {
    CryptoPP::ByteQueue out = Elonef::toBytes(this->sign_key);
    Elonef::toBytes(data_key).TransferAllTo(out);
    Elonef::toBytes(server_key).TransferAllTo(out);
    Elonef::toBytes(ca_keys.begin(), ca_keys.end()).TransferAllTo(out);
    Elonef::toBytes(uid).TransferAllTo(out);   
    return out;
}

Elonef::KeyPair Elonef::generateKeyPairRSA() {
    CryptoPP::InvertibleRSAFunction params;
    params.GenerateRandomWithKeySize(Elonef::rng, 3072);

    CryptoPP::RSA::PrivateKey privateKey(params);
    CryptoPP::RSA::PublicKey publicKey(params);

    Elonef::KeyPair out;
    {
        CryptoPP::StringSink sink(out.publicKey);
        CryptoPP::PEM_Save(sink, publicKey);
    };
    {
        CryptoPP::StringSink sink(out.privateKey);
        CryptoPP::PEM_Save(sink, privateKey);
    };

    return out;
}

Elonef::KeyPair Elonef::generateKeyPairECDSA() {
    Elonef::ECDSA::PrivateKey priv;
    Elonef::ECDSA::PublicKey  pub;

    priv.Initialize( Elonef::rng, CryptoPP::ASN1::secp256k1() );
    CryptoPP::ByteQueue privateKey;
    priv.Save(privateKey);

    priv.MakePublicKey(pub);
    CryptoPP::ByteQueue publicKey;
    pub.Save(publicKey);

    Elonef::KeyPair keyPair;
    keyPair.privateKey = toBase64(privateKey);
    keyPair.publicKey  = toBase64(publicKey);

    return keyPair;    
}

Elonef::ClientKeys Elonef::generateClientKeys(std::string server_key, std::unordered_map<std::string, std::string> ca_keys, std::string id) {
    KeyPair ecdsa_keys = generateKeyPairECDSA();
    KeyPair rsa_key = generateKeyPairRSA();

    ClientKeys out;
    out.private_key.data_key   = rsa_key.   privateKey;
    out.private_key.sign_key   = ecdsa_keys.privateKey;
    out.private_key.server_key = server_key;
    out.private_key.ca_keys    = ca_keys;
    out.private_key.uid        = id;

    out.public_key.data_key.key = rsa_key.   publicKey;
    out.public_key.sign_key.key = ecdsa_keys.publicKey;
    out.public_key.id           = id;
    return out;
}